library/alloc/src/alloc.rs

```diff
@@ -90,7 +93,14 @@ pub use std::alloc::Global;
 #[must_use = "losing the pointer will leak memory"]
 #[inline]
 pub unsafe fn alloc(layout: Layout) -> *mut u8 {
-    unsafe { __rust_alloc(layout.size(), layout.align()) }
+    unsafe {
+        // Make sure we don't accidentally allow omitting the allocator shim in
+        // stable code until it is actually stabilized.
+        #[cfg(not(bootstrap))]
+        core::ptr::read_volatile(&__rust_no_alloc_shim_is_unstable);
```

Comment:
I did expect the perf regression to come from this change. It is a single extra instruction on the allocation path to ensure `__rust_no_alloc_shim_is_unstable` must be defined if no allocator shim is linked in. The only way I can think of that guarantees that it isn't possible to link without defining `__rust_no_alloc_shim_is_unstable` would be to put `__rust_alloc` and an item referencing `__rust_no_alloc_shim_is_unstable` in the same COMDAT group, but this isn't possible for all object file formats and rust doesn't have a way to do this without `global_asm!()`.

Reply:
It's unfortunate that this symbol is now present in every single allocation path, especially where it enlarges the output binary for platforms like Wasm.

Reply:
That would prevent LLVM from optimizing allocations away I think as `__rust_no_alloc_shim_is_unstable` is not considered to be an allocator function by LLVM, while `__rust_alloc` is. That said, I just opened a draft PR which will enable doing away with the allocator shim entirely in the future, after which I hope it will be much easier to request stabilization of support for not using the allocator shim and thus remove this symbol entirely.